skimage/filters/_sparse.py

```diff
@@ -0,0 +1,99 @@
+import numpy as np
+from ._sparse_cy import _correlate_sparse_offsets
+from collections.abc import Iterable
+
+
+def _validate_window_size(axis_sizes):
+    """Ensure all sizes in ``axis_sizes`` are odd.
+
+    Parameters
+    ----------
+    axis_sizes : iterable of int
+
+    Raises
+    ------
+    ValueError
+        If any given axis size is even.
+    """
+    for axis_size in axis_sizes:
+        if axis_size % 2 == 0:
+            msg = ('Window size for `threshold_sauvola` or '
+                   '`threshold_niblack` must not be even on any dimension. '
+                   'Got {}'.format(axis_sizes))
+            raise ValueError(msg)
+
+
+def _to_np_mode(mode):
+    """Convert padding modes from `ndi.correlate` to `np.pad`."""
+    mode_translation_dict = dict(nearest='edge', reflect='symmetric',
+                                 mirror='reflect')
+    if mode in mode_translation_dict:
+        mode = mode_translation_dict[mode]
+    return mode
+
+
+def correlate_sparse(image, kernel, mode='reflect'):
+    """Compute valid cross-correlation of `padded_array` and `kernel`.
+
+    This function is *fast* when `kernel` is large with many zeros.
+
+    See ``scipy.ndimage.correlate`` for a description of cross-correlation.
+
+    Parameters
+    ----------
+    image : ndarray, dtype float, shape (M, N,[ ...,] P)
+        The input array. If mode is 'valid', this array should already be
+        padded, as a margin of the same shape as kernel will be stripped
+        off.
+    kernel : ndarray, dtype float shape (Q, R,[ ...,] S)
+        The kernel to be correlated. Must have the same number of
+        dimensions as `padded_array`. For high performance, it should
+        be sparse (few nonzero entries).
+    mode : string, optional
+        See `scipy.ndimage.correlate` for valid modes. 
+        Additionally, mode 'valid' is accepted, in which case no padding is 
+        applied and the result is the result for the smaller image for which 
+        the kernel is entirely inside the original data.
+
+    Returns
+    -------
+    result : array of float, shape (M, N,[ ...,] P)
+        The result of cross-correlating `image` with `kernel`. If mode
+        'valid' is used, the resulting shape is (M-Q+1, N-R+1,[ ...,] P-S+1).
+    """
+    if not isinstance(kernel, np.ndarray):
+        msg = '`correlate_sparse` kernel must be an numpy array object.'
+        raise ValueError(msg)
+
+    if mode == 'valid':
```

Comment:
@rfezzani afaik there is no mode name `valid` in `ndi.correlate` we have this specific mode check here.

Reply:
Oh, I definitely missed something :sweat_smile:...
This is may be opinionated, but I vote for removing this option! It comes from [`scipy.signal.convolve`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.convolve.html), and I personally  don't like it: a user needing the valid part only have to crop the output by him self!

Reply:
@rfezzani in this case it is incredibly useful because we don't want to pad the image if the user has already padded it. Without it, we *have* to pad.